src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,459 @@
+                self._raise_api_error("Failed to fetch workspaces!", api_call)
+                return False
+            df = pd.DataFrame(api_call.json()["value"], columns=["id", "name"])
+
+            if self.workspace_name is None:
+                self._show_workspaces(df)
+                return False
+
+            rows = df.loc[df["name"] == self.workspace_name, "id"]
+            if rows.empty:
+                self._raise_error(
+                    f"Workspace name '{self.workspace_name}' cannot be found!"
+                )
+                return False
+            self.workspace_id = rows.values[0]
+
+        return True
+
+    def _get_dataset(self) -> bool:
+        """
+        Gets dataset ID based on the dataset name, or shows all datasets
+            if no parameter was specified.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if self.dataset_id is None:
+            api_call = requests.get(
+                url=f"{self.powerbi_url}groups/{self.workspace_id}/datasets",
+                headers=self.api_header,
+            )
+            if api_call.status_code != 200:
+                self._raise_api_error("Failed to fetch datasets!", api_call)
+                return False
+            df = pd.DataFrame(api_call.json()["value"], columns=["id", "name"])
+
+            if self.dataset_name is None:
+                self._show_datasets(df)
+                return False
+
+            rows = df.loc[df["name"] == self.dataset_name, "id"]
+            if rows.empty:
+                self._raise_error(
+                    f"Dataset name '{self.dataset_name}' cannot be found, "
+                    "or the dataset doesn't have a user with the required permissions!"
+                )
+                return False
+            self.dataset_id = rows.values[0]
+
+        return True
+
+    def _connect(self) -> bool:
+        """
+        Connects or reconnects to PowerBI to fetch refresh history or trigger a refresh.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if not self._get_access_token():
+            return False
+        if not self._get_workspace():
+            return False
+        if not self._get_dataset():
+            return False
+
+        return True
+
+    def _get_last_refresh(self) -> bool:
+        """
+        Gets the latest record in the PowerBI dataset refresh history.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if not self._connect():
+            return False
+
+        api_url = (
+            f"{self.powerbi_url}groups/{self.workspace_id}"
```

Comment:
FYI: Lets say the workspace_name is defined, but the workspace_id is None, this api_call fails.